internal/find/finder.go

```diff
@@ -0,0 +1,259 @@
+		if err != nil {
+			return err
+		}
+		if dirEntry.Type() != 0 {
+			return nil
+		}
+		fileInfo, err := dirEntry.Info()
+		if err != nil {
+			return err
+		}
+		size := fileInfo.Size()
+		Stats.TotalBytes.Add(uint64(size))
+		regularFilesCh <- pathWithSize{
+			path: path,
+			size: size,
+		}
+		return nil
+	}
+	concurrentWalkDir(root, errChan, walkDirFunc)
+}
+
+// findPathsWithIdenticalSizes reads paths from regularFilesCh and, once there
+// are more than threshold paths with the same size, writes them to
+// pathsToHashCh.
+func findPathsWithIdenticalSizes(pathsToHashCh chan<- pathWithSize, regularFilesCh <-chan pathWithSize, threshold int) {
+	allPathsBySize := make(map[int64][]pathWithSize)
+	for pathWithSize := range regularFilesCh {
+		pathsBySize := append(allPathsBySize[pathWithSize.size], pathWithSize) //nolint:gocritic
+		allPathsBySize[pathWithSize.size] = pathsBySize
+		if len(pathsBySize) == threshold {
+			for _, p := range pathsBySize {
+				pathsToHashCh <- p
+			}
+		} else if len(pathsBySize) > threshold {
+			pathsToHashCh <- pathWithSize
+		}
+	}
+}
+
+// hashPaths reads paths from pathsToHashCh, computes their hashes, and writes
+// them to pathsWithHashCh.
+func hashPaths(pathsToHashCh <-chan pathWithSize, pathsWithHashCh chan<- pathWithHash, errChan chan<- error) {
+	var wg sync.WaitGroup
+	for pathWithSize := range pathsToHashCh {
+		pathWithSize := pathWithSize
+		wg.Add(1)
+		_ = ants.Submit(func() {
+			pathWithHash, err := pathWithSize.pathWithHash()
+			if err != nil {
+				errChan <- err
+			} else {
+				pathsWithHashCh <- pathWithHash
+			}
+			wg.Done()
+		})
+	}
+	wg.Wait()
+}
+
+// pathWithHash hashes p and returns a pathWithHash.
+func (p pathWithSize) pathWithHash() (pathWithHash, error) {
+	hash, err := p.hash()
+	if err != nil {
+		return pathWithHash{}, err
+	}
+	pathWithHash := pathWithHash{
+		path: p.path,
+		hash: hash,
+	}
+	return pathWithHash, nil
+}
+
+// hash returns p's hash.
+func (p pathWithSize) hash() (hash, error) {
+	if p.size == 0 {
+		return xx3HashSumZero, nil
+	}
+	file, err := os.Open(p.path)
+	if err != nil {
+		return xxh3.Uint128{}, err
+	}
+	Stats.FilesOpened.Add(1)
+	defer file.Close()
+	hash := xxh3.New()
+	written, err := io.Copy(hash, file)
+	if err != nil {
+		return xxh3.Uint128{}, err
+	}
+	Stats.BytesHashed.Add(uint64(written))
+	return hash.Sum128(), nil
+}
+
+func (f *Finder) FindDuplicates() (map[string][]string, error) {
+	defer ants.Release()
+
+	var errHandler func(error) error
+	if f.KeepGoing {
+		errHandler = func(err error) error {
+			if err != nil {
+				Stats.Errors.Add(1)
+				fmt.Fprintln(os.Stderr, err)
+			}
+			return nil
+		}
+	} else {
+		errHandler = func(err error) error {
+			return err
+		}
+	}
+
+	errChan := make(chan error, channelBufferCapacity)
+	defer close(errChan)
+
+	// Generate paths with size.
+	regularFilesCh := make(chan pathWithSize, channelBufferCapacity)
+	_ = ants.Submit(func() {
```

Comment:
Note that we shouldn't use `ants.Submit` for these tasks because it is essential that they always run. If these tasks get queued then the channels can block. I suspect that the code happens to work now because these tasks are created before others, and therefore happen to be scheduled on the initial workers.